app/src/main/java/fi/bitrite/android/ws/ui/MapFragment.java

```diff
@@ -341,95 +405,101 @@ private void doInitialMapMove() {
-                                        mClusterManager.addItem(ClusterUser.from(user));
-                                    }
+                                    addUserToCluster(user.toSimpleUser());
                                 }
-                                mClusterManager.cluster();
+                                mMarkerClusterer.invalidate();
+                                mMap.invalidate();
                             }, throwable -> {
                                 // TODO(saemy): Error handling.
                                 Log.e(TAG, throwable.getMessage());
                             }));
         }
+
     }
 
     private void loadOfflineUsers() {
-        mClusterManager.clearItems();
-        mClusterManager.getMarkerCollection().clear();
-        mClusteredUsers.clear();
-
-        // We'll use the starred users when network is offline.
-        List<Integer> loadedUserIds = new ArrayList<>();
+        // We'll show the starred users until we load a fresh version of them.
         mLoadOfflineUserDisposable = Observable.merge(mFavoriteRepository.getFavorites())
+                .filter(Resource::hasData)
+                .map(userResource -> userResource.data)
                 .observeOn(AndroidSchedulers.mainThread())
-                .subscribe(userResource -> {
-                    if (userResource.hasData() && mIsOffline) {
-                        // Users pop up twice as one is the error since we cannot load it from the
-                        // network.
-                        User user = userResource.data;
-                        if (loadedUserIds.contains(user.id)) {
-                            return;
-                        }
-                        loadedUserIds.add(user.id);
-
-                        mClusterManager.addItem(ClusterUser.from(user));
-                        mClusterManager.cluster();
+                .subscribe(user -> {
+                    // Users pop up twice as one is the error since we might not be able to load it
+                    // from the network.
+                    boolean added = mOfflineUserIds.add(user.id);
+                    if (!added) {
+                        return;
                     }
+
+                    addUserToCluster(user);
+                    mMarkerClusterer.invalidate();
                 });
-        getResumePauseDisposable().add(mLoadOfflineUserDisposable);
+        getCreateDestroyDisposable().add(mLoadOfflineUserDisposable);
+    }
+
+    private void addUserToCluster(SimpleUser user) {
+        // Only add to the cluster if it wasn't before or when its location changed.
+        final Marker existingMarker = mClusteredUsers.get(user.id);
+        boolean isNew = existingMarker == null
+                        || !existingMarker.getPosition().equals(user.location);
+        if (isNew) {
+            if (existingMarker != null) {
+                // FIXME(saemy): Remove the old marker.
```

Comment:
Shouldn’t this be done first before we are merging this PR?

Reply:
Done.